pydis_site/apps/api/migrations/0044_migrate_nominations_from_infraction_to_nomination_model.py

```diff
@@ -0,0 +1,64 @@
+# Generated by Django 2.2.5 on 2019-09-30 12:15
+import logging
+
+from django.db import migrations
+from django.db.models import Q
+
+log = logging.getLogger('nomination_migration')
+
+
+def migrate_nominations_to_new_model(apps, schema_editor):
+    """
+    Migrations nominations from the infraction model to the nomination model.
+
+    This migration works by replaying the nomination history in chronological order, adding and
+    ending nominations as we've recorded them.
+    """
+    Infraction = apps.get_model('api', 'Infraction')
+    Nomination = apps.get_model('api', 'Nomination')
+
+    all_nominations = (
+        Q(reason__startswith="Helper nomination:") | Q(reason__startswith="Unwatched (talent-pool):")
+    )
+
+    for infraction in Infraction.objects.filter(all_nominations).order_by('inserted_at'):
+        if infraction.reason.startswith("Helper nomination:"):
+            if Nomination.objects.filter(user=infraction.user, active=True).exists():
```

Comment:
Would this situation arise from this?

> In addition, some nominations may not have received an end_reason when ending the nomination; these nominations will appear as active incorrectly.

Reply:
On second though I don't think so since they wouldn't even have matched the reason filter, right?

Reply:
> Would this situation arise from this?
> 
> > In addition, some nominations may not have received an end_reason when ending the nomination; these nominations will appear as active incorrectly.

No, it does not arise from it. Currently, if someone has received a first nomination that was ended before we recorded "end reasons" and then received a new nomination, the new nomination would be rejected, since someone can't have two active nominations at the same time. If we'd then observe an "end reason", the first nomination would be ended with that reason. We'd end up in the correct *state* (no active nomination), but with some loss of intermediate information.

It's a bit of an edge case and it does not happen in the historical data we have, but another solution would be to close the first active nomination when we observe a new one with a `"(no reason recorded)"` and then add the second one, which means the end reason we then observe will matched to the correct reason. 

This has another edge case that may occur, though, and one that actually did happen in our historical data: For a brief period of time, if two people nominated someone more or less simultaneously, the second nomination would be recorded in the same way as the first (as a new active nomination), with no end reason between those two. With the above "solution", this would make it seem as if the individual had been nominated multiple times, being rejected for the first nominations, while in reality it was just one "nomination period" and multiple people were excited about them. (This mainly happened when a helper nominated someone that was already in talent-pool, since they had no way of knowing that.)

Ultimately, though, I've checked all historical data and we're talking about one or two individuals who's nominations resulted in the Helper role and the current migration just makes it look like they were nomination once and the end reason spells out they were accepted and offered the role. The only thing we're missing out on is that a second individual also nominated them during their nomination period.

Reply:
Do I understand correctly?

* If `"(no reason recorded)"` approach is used, it may end up in keeping a duplicate start nomination because of that bug you described with simultaneously nominating someone.
* If the current approach is used, it can result in a skew as you described in the other comment, but in practice, this won't happen given our historical data
* This check for if an active infraction already exists is not just there as a safety measure but actually has a use for the simultaneous nomination bug to discard the duplicate.

Reply:
Yes, each additional nomination while the first is still active would instead appear as a separate new nomination. For the two users this is relevant, it would appear as if they were nominated on three or four separate occasions instead of the one nomination period both actually have had.